Tools/TestWebKitAPI/Tests/mac/SkipAdInPictureInPicture.mm

```diff
@@ -0,0 +1,144 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+
+#if HAVE(PIP_SKIP_PREROLL)
+
+#import "PlatformUtilities.h"
+#import "Test.h"
+#import "TestWKWebView.h"
+#import "WKWebViewConfigurationExtras.h"
+#import <WebKit/WKPreferencesPrivate.h>
+#import <WebKit/WKUIDelegatePrivate.h>
+#import <WebKit/WKWebViewPrivate.h>
+#import <pal/spi/mac/PIPSPI.h>
+#import <wtf/RetainPtr.h>
+#import <wtf/SoftLinking.h>
+#import <wtf/cocoa/TypeCastsCocoa.h>
+
+SOFT_LINK_PRIVATE_FRAMEWORK(PIP)
+SOFT_LINK_CLASS(PIP, PIPPanel)
+SOFT_LINK_CLASS(PIP, PIPPrerollAttributes)
+SOFT_LINK_CLASS(PIP, PIPViewController)
+
+static bool didEnterPiP;
+
+@interface SkipAdInPiPFullscreenDelegate : NSObject <WKUIDelegate>
+@end
+
+@implementation SkipAdInPiPFullscreenDelegate
+
+- (void)_webView:(WKWebView *)webView hasVideoInPictureInPictureDidChange:(BOOL)value
+{
+    if (value)
+        didEnterPiP = true;
+}
+
+@end
+
+namespace TestWebKitAPI {
+
+static RetainPtr<TestWKWebView> createWebView()
+{
+    RetainPtr configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"WebProcessPlugInWithInternals" configureJSCForTesting:YES];
+    [configuration preferences]._allowsPictureInPictureMediaPlayback = YES;
+    [configuration setMediaTypesRequiringUserActionForPlayback:WKAudiovisualMediaTypeAudio];
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 300, 300) configuration:configuration.get() addToWindow:YES]);
+    return webView;
+}
+
+TEST(SkipAdInPictureInPicture, VideoHasAd)
+{
+    auto webView = createWebView();
+    RetainPtr<SkipAdInPiPFullscreenDelegate> handler = adoptNS([[SkipAdInPiPFullscreenDelegate alloc] init]);
+    [webView setUIDelegate:handler.get()];
+
+    [webView synchronouslyLoadTestPageNamed:@"SkipAdInPictureInPicture"];
+
+    didEnterPiP = false;
+    [webView evaluateJavaScript:@"document.getElementById('enter-pip').click()" completionHandler:nil];
+    TestWebKitAPI::Util::waitFor([&] {
+        return didEnterPiP;
+    });
+    EXPECT_TRUE(didEnterPiP);
+
+    __block NSWindow *pipPanel = nil;
+    [[NSApplication sharedApplication] enumerateWindowsWithOptions:0 usingBlock:^(NSWindow *window, BOOL *stop) {
+        if ([window isKindOfClass:getPIPPanelClass()]) {
+            pipPanel = window;
+            *stop = YES;
+        }
+    }];
+    ASSERT_TRUE(pipPanel);
+    auto *contentViewController = [pipPanel contentViewController];
```

Comment:
Not the most important for testing code, but it's a good habit to start using `RetainPtr` everywhere instead of raw pointers.